### ISSUE TITLE
[HOP] Allow torch.cond (and other HOP) to allow returning alias and inplace mutation when run under `torch.no_grad` or `torch.inference_mode`

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#172325]
### 🚀 The feature, motivation and pitch

Fixing this should help unblock #168911 

Motivations (as mentioned in #168911)

> Some motivation for this:
> 
> Firstly, the desire to do basic data-dependent computation has come up several times, but one of the key use cases is skipping the optimizer in the case of a NaN gradient in mixed precision training.
> 
> Secondly, figuring out making torch.cond() work with cuda graphs paves the way for adding support for torch.while_loop() in cuda graphs, which may be used to do MoE inside a cuda graph without dropping tokens, but that is more complicated and requires its own issue to write down those issues (they go beyond just while_loop support for cuda graphs; it may also need tricky memory management for autograd).

Repro that should work with the fix:

```python
import pytest
import torch

def test_cond_no_grad():
    predicate_true = torch.tensor(True, device="cuda")
    predicate_false = torch.tensor(False, device="cuda")
    org_data = torch.ones(1024, device="cuda")

    def fn(predicate, data):
        return torch.cond(predicate, lambda x: x, lambda x: x.sin_().add_(2), [data])

    def _test(fn):
        with torch.no_grad():
            expected = org_data.sin() + 2
            data = org_data.clone()
            output = fn(predicate_false, data)
            torch.testing.assert_close(output, expected)
            assert id(output) == id(data)

            data = org_data.clone()
            output = fn(predicate_true, data)
            torch.testing.assert_close(output, org_data)
            assert id(output) == id(data)

    # Exception expected as `torch.is_grad_enabled()` returns True.
    with pytest.raises(torch._dynamo.exc.UncapturedHigherOrderOpError):
        fn(predicate_true, org_data)

    _test(fn)

    cfn = torch.compile(fn)

    with pytest.raises(torch._dynamo.exc.UncapturedHigherOrderOpError):
        cfn(predicate_true, org_data)

    _test(cfn)

    with pytest.raises(torch._dynamo.exc.UncapturedHigherOrderOpError):
        cfn(predicate_true, org_data)

    _test(cfn)
```

cc: @kiya00

### Additional context

Potential Patch could look like this - https://github.com/pytorch/pytorch/issues/168911#issuecomment-3641954565

cc: @zou3519 @ydwu4 for your thoughts

cc @chauhang @penguinwu @ydwu4 @bdhirsh @bobrenjc93

**Comment by kshitij12345** [comment]
Thanks for the confirmation @ydwu4 , my colleague @kiya00 will start working on this.
